File: services/client/src/protocol/communication/communication.go
package communication

import (
	"net"
	"time"

	common "github.com/7574-sistemas-distribuidos/tp-nivelador/src/protocol/common"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

// Envia un Packet a traves de la conexion. Primero envia el largo del paquete en bytes,
// y posteriormente los bytes del paquete en si.
func SendPacket(socket net.Conn, packet Packet) error {
	if err := socket.SetWriteDeadline(time.Now().Add(common.SOCKET_WRITE_TIMEOUT_SEC * time.Second)); err != nil {
		return err
	}

	packetBytes := packet.ToBytes()
	packetLength := uint16(len(packetBytes))
	packetLengthBytes := common.Uint16ToBytes(packetLength)

	fullPacket := append(packetLengthBytes, packetBytes...)
	if err := safe_socket.SendAll(socket, fullPacket); err != nil {
		return err
	}
	return nil
}

// Recibe un Packet a traves de la conexion. Primero lee los bytes del largo del paquete (fijo),
// y posteriormente, utilizado el dato del largo, lee el paquete entero.
func ReceivePacket(socket net.Conn) (Packet, error) {
	if err := socket.SetReadDeadline(time.Now().Add(common.SOCKET_READ_TIMEOUT_SEC * time.Second)); err != nil {
		return Packet{}, err
	}

	packetLengthBytes, err := safe_socket.RecvAll(socket, PACKET_LENGTH_BYTES)
	if err != nil {
		return Packet{}, err
	}
	packetLength, err := common.BytesToUint16(packetLengthBytes)
	if err != nil {
		return Packet{}, err
	}

	packetBytes, err := safe_socket.RecvAll(socket, int(packetLength))
	if err != nil {
		return Packet{}, err
	}
	packet, err := PacketFromBytes(packetBytes)
	if err != nil {
		return Packet{}, err
	}

	return packet, nil
}

File: services/server/src/protocol/communication/communication.py
from protocol.common.utils import uint16_to_bytes, bytes_to_uint16
import safe_socket
from .packet import Packet, PACKET_LENGTH_BYTES

# Envia un Packet a traves de la conexion. Primero envia el largo del paquete en bytes,
# y posteriormente los bytes del paquete en si.
def send_packet(socket, packet):
    packet_bytes = packet.to_bytes()
    packet_length = len(packet_bytes)
    packet_length_bytes = uint16_to_bytes(packet_length)

    full_packet = packet_length_bytes + packet_bytes
    safe_socket.send_all(socket, full_packet)


# Recibe un Packet a traves de la conexion. Primero lee los bytes del largo del paquete (fijo),
# y posteriormente, utilizado el dato del largo, lee el paquete entero.
def receive_packet(socket) -> Packet:
    packet_length_bytes = safe_socket.recv_all(socket, PACKET_LENGTH_BYTES)
    packet_length = bytes_to_uint16(packet_length_bytes)
    packet_bytes = safe_socket.recv_all(socket, packet_length)

    return Packet.from_bytes(packet_bytes)